cluster/src/main/scala/org/apache/spark/memory/SnappyUnifiedMemoryManager.scala

```diff
@@ -535,28 +568,34 @@ class SnappyUnifiedMemoryManager private[memory](
   override def releaseStorageMemoryForObject(objectName: String,
                                              numBytes: Long,
                                              memoryMode: MemoryMode): Unit = synchronized {
-    logDebug(s"releasing [SNAP] memory for $objectName = $numBytes")
-    if (memoryForObject.addTo(objectName, -numBytes) >= 0L) {
+    logDebug(s"releasing $managerId memory for $objectName = $numBytes")
+    if (memoryForObject.addTo(objectName -> memoryMode, -numBytes) != 0L) {
       super.releaseStorageMemory(numBytes, memoryMode)
     } else {
       // objectName was not present
       memoryForObject.removeLong(objectName)
     }
   }
 
-  override def releaseStorageMemory(numBytes: Long, memoryMode: MemoryMode): Unit =
+  override def releaseStorageMemory(numBytes: Long, memoryMode: MemoryMode): Unit = {
     releaseStorageMemoryForObject(SPARK_CACHE, numBytes, memoryMode)
+    logDebug(s"releasing $managerId memory for $SPARK_CACHE $numBytes")
+    if (memoryForObject.containsKey(SPARK_CACHE)) {
+      memoryForObject.addTo(SPARK_CACHE -> memoryMode, -numBytes)
```

Comment:
Key in containsKey is incorrect -- should be a pair with memoryMode. Better to create the key once and use in both.

A more efficient version that will avoid two key lookups for most cases:

```scala
val key = SPARK_CACHE -> memoryMode
if (memoryForObject.addTo(key, -numBytes) != 0L) {
  super.releaseStorageMemory(numBytes, memoryMode)
} else {
  // key was not present so remove
  memoryForObject.removeLong(key)
}
```